idaes/core/solvers/petsc.py

```diff
@@ -0,0 +1,680 @@
+                parent = var.parent_block()
+                name = var.local_name + "_disc_eq"
+                disc_eq = getattr(parent, name)
+                disc_eqns.append(disc_eq)
+    return disc_eqns
+
+
+def _set_dae_suffixes_from_variables(m, variables, deriv_diff_map):
+    """Write suffixes used by the solver to identify different variable types
+    and associated derivative and differential variables.
+
+    Args:
+        m: model to search for variables and write suffixes to
+        variables (list): List of time indexed variables at a specific time
+            point
+        deriv_diff_map (ComponentMap): Maps DerivativeVar data objects to
+            differential variable data objects
+
+    Returns:
+        None
+    """
+    # The dae_suffix provides the solver information about variables types
+    # algebraic, differential, derivative, or time, see DaeVarTypes
+    m.dae_suffix = pyo.Suffix(
+        direction=pyo.Suffix.EXPORT,
+        datatype=pyo.Suffix.INT,
+    )
+    # The dae_link suffix provides the solver a link between the differential
+    # and derivative variable, by assigning the pair a unique integer index.
+    m.dae_link = pyo.Suffix(
+        direction=pyo.Suffix.EXPORT,
+        datatype=pyo.Suffix.INT,
+    )
+    dae_var_link_index = 1
+    differential_vars = []
+    i = 0
+    for var in variables:
+        if var in deriv_diff_map:
+            deriv = var
+            diffvar = deriv_diff_map[deriv]
+            m.dae_suffix[diffvar] = int(DaeVarTypes.DIFFERENTIAL)
+            m.dae_suffix[deriv] = int(DaeVarTypes.DERIVATIVE)
+            m.dae_link[diffvar] = dae_var_link_index
+            m.dae_link[deriv] = dae_var_link_index
+            i += 1
+            dae_var_link_index += 1
+            if not diffvar.fixed:
+                differential_vars.append(diffvar)
+    return differential_vars
+
+
+def _get_derivative_differential_data_map(m, time):
+    """Get a map from data objects of derivative variables to the
+    corresponding data objects of differential variables.
+
+    Args:
+        m: Model to search for DerivativeVars
+        time: Set with respect to which DerivativeVars must be differentiated
+
+    Returns:
+        (ComponentMap): Map from derivative data objects to differential
+            data objects
+    """
+    derivs = ComponentSet()
+    for con in m.component_data_objects(pyo.Constraint, active=True):
+        for var in identify_variables(con.body):
+            pvar = var.parent_component()
+            if (isinstance(pvar, pyodae.DerivativeVar) and not var.fixed and
+                    time in ComponentSet(pvar.get_continuousset_list())):
+                derivs.add(var)
+    deriv_diff_list = [None]*len(derivs)
+    for i, deriv in enumerate(derivs):
+        diffvar = deriv.parent_component().get_state_var()
+        idx = deriv.index()
+        deriv_diff_list[i] = (deriv, diffvar[idx])
+    return ComponentMap(deriv_diff_list)
```

Comment:
I'd like to avoid calling `deriv.index()` here if we can. This is a linear scan over `deriv`'s indexing set, and so will be quite expensive if we are calling this for every derivative data object. See my PR into this branch, https://github.com/eslickj/idaes-pse/pull/2, where I propose an alternative implementation.